### PR TITLE
HSEARCH-295 Automatically determining filter key based on all parameter names and values

### PR DESCRIPTION
**Pull request by gunnarmorling**
* The `@Key` method is optional now; By default a `StandardFilterKey` will be used which is based on all presented parameter names and values
* Avoiding filter (factory) instantiation if not needed

@Sanne Is that what you had in mind for [HSEARCH-295](https://hibernate.atlassian.net/browse/HSEARCH-295)?

### CODE REVIEW

**Comment by gunnarmorling** [comment]
@Sanne re-worked some of the commits; So best to re-pull on your side.

**Comment by gunnarmorling** [comment]
Pushed one more commit for the optimized filter key implementation and changed one commit message to refer to HSEARCH-818.

Waiting a bit for feedback to your mail, then I'll do the suggested deprecation / doc removal.

**Comment by Sanne** [comment]
@Hibernate-CI retest this please

**Comment by Sanne** [comment]
I actually realized this PR doesn't need to wait for a consensus on the deprecation decision; I've separated that into a separate issue: https://hibernate.atlassian.net/browse/HSEARCH-1785

**Comment by gunnarmorling** [comment]
Closing as already merged by @Sanne.

File: engine/src/main/java/org/hibernate/search/annotations/FullTextFilterDef.java [hunk 1]
```diff
@@ -23,21 +23,30 @@
...
 	/**
-	 * Either implements {@link org.apache.lucene.search.Filter}
-	 * or contains a <code>@Factory</code> method returning one.
-	 * The generated <code>Filter</code> must be thread-safe.
-	 *
-	 * If the filter accept parameters, an <code>@Key</code> method must be present as well.
-	 *
-	 * @return a class which either implements <code>Filter</code> directly or contains a method annotated with
-	 * <code>@Factory</code>.
+	 * The implementation of this filter definition. May be
```
Comment (hferentschik) [inline engine/src/main/java/org/hibernate/search/annotations/FullTextFilterDef.java]: nice docs :-)

**Pull request closed without merging**